### DISCUSSION TITLE
LoadError: Method overwriting is not permitted during Module precompile

### DISCUSSION THREAD

**Discussion by singularitti** [JuliaLang/julia#50725] [General]
I'm seeing a `ERROR: LoadError: Method overwriting is not permitted during Module precompile`. on Julia v1.10.0-beta1 when precompiling my package. It did not happen on v1.9. Is there a behavioral change that was not mentioned in [NEWS.md](https://github.com/JuliaLang/julia/blob/v1.10.0-beta1/NEWS.md)?

**Comment by vtjnash** [comment]
It has always given a fatal error warning, so it is not a news change. But recently things got more crash-y for people that missed the warning (previously it continued so you could see all of the warnings at once, but now it just exits immediately at the first code mistake), so we made it fail more clearly, to help code authors see that more easily on CI runs.

**Reply by singularitti** [reply]
That means in the future versions of Julia, method overwriting will be banned? Or is this just a temporary change?

**Reply by vtjnash** [reply]
It has always been banned, during precompile. It is fine during runtime, and is the basis for tools like Revise.jl

**Reply by singularitti** [reply]
Due to some reasons, [my code](https://github.com/singularitti/Thinkers.jl/blob/3cea7d23687f8097712eee6a603eb1cd2a30bbb9/src/TimeLimitedThunk.jl#L8-L18) needs this dynamic overwriting feature. And I see no easy way to avoid this without changing my API. Now I cannot load this package on v1.10.0.

**Reply by vtjnash** [reply]
It looks like you just have the code in the wrong place. There are several ways to change default constructors. For example, they can all be put inside the struct definition, like so:
```julia
struct TimeLimitedThunk <: WrappedThink
           time_limit::Period
           wrapped::Thunk
       TimeLimitedThunk(time_limit, callable, args::Tuple, kwargs::Iterators.Pairs) =
           new(time_limit, Thunk(callable, args, kwargs))
       TimeLimitedThunk(time_limit, callable; kwargs...) =
           new(time_limit, Thunk(callable; kwargs...))
       TimeLimitedThunk(time_limit, callable, arg, args...; kwargs...) =
           new(time_limit, Thunk(callable, arg, args...; kwargs...))

       TimeLimitedThunk(time_limit::Period, callable::Thunk) =
           new(time_limit, callable)
end
```

**Reply by singularitti** [reply]
You're right. I forgot to put them inside. So back to my original question, will the failing behavior continue in the official release of v1.10.0, or will it be reverted back to the behavior of v1.9?

**Reply by vtjnash** [reply]
It also is a failure on v1.9 and earlier

**Reply by singularitti** [reply]
It is. But I could load my package without error on v1.9 and earlier.

**Reply by timholy** [reply]
The package ecosystem can [become completely broken by method overwriting](https://discourse.julialang.org/t/what-is-incremental-compilation-and-what-does-it-mean-for-it-to-be-broken/100956/2?u=tim.holy), so this indeed should be an error. You should make the changes to your code.

**Reply by vtjnash** [reply]
It did error in v1.9 too, you just didn't read or ignored it.

**Reply by singularitti** [reply]
It did error, but I was still able to load the package.
